miasm2/core/objc.py

```diff
@@ -0,0 +1,1712 @@
+        obj = ast_gen_struct_align_size_compil(ast, types_mngr)
+    elif types_mngr.layout == types_mngr.LAYOUT_PACKED:
+        obj = ast_gen_struct_align_size_packed(ast, types_mngr)
+    else:
+        raise RuntimeError("Unknown layout %r" % types_mngr.layout)
+
+    return obj
+
+
+def ast_gen_union_align_size_compil(ast, types_mngr):
+    """Evaluates the size/alignment of a C ast union
+    (default packed)
+
+    @ast: parsed pycparser.c_ast object
+    @types_mngr: types manager
+    """
+
+    offset = 0
+
+    align_max, size_max = 0, 0
+
+    if ast.name is None:
+        name = types_mngr.gen_temp_name()
+    else:
+        name = ast.name
+    new_obj = ObjCUnion(name)
+
+    for arg in ast.decls:
+        obj = ast_get_align_size(arg, types_mngr)
+        align_max = max(align_max, obj.align)
+        size_max = max(size_max, obj.size)
+        new_obj.add_field(arg.name, obj,
+                          offset, obj.size)
+
+    new_obj.set_align_size(align_max, size_max)
+    types_mngr.add_type((name, ), new_obj)
+    return new_obj
+
+
+def ast_gen_union_align_size_packed(ast, types_mngr):
+    """Evaluates the size/alignment of a C ast union
+    (packed)
+
+    @ast: parsed pycparser.c_ast object
+    @types_mngr: types manager
+    """
+
+    offset = 0
+
+    align_max, size_max = 0, 0
+
+    if ast.name is None:
+        name = types_mngr.gen_temp_name()
+    else:
+        name = ast.name
+    new_obj = ObjCUnion(name)
+
+    for arg in ast.decls:
+        obj = ast_get_align_size(arg, types_mngr)
+        align_max = 1
+        size_max = max(size_max, obj.size)
+        new_obj.add_field(arg.name, obj,
+                          offset, obj.size)
+
+    new_obj.set_align_size(align_max, size_max)
+    types_mngr.add_type((name, ), new_obj)
+    return new_obj
+
+
+def ast_gen_union_align_size(ast, types_mngr):
+    """Evaluates the size/alignment of a C ast union
+    (packed style in type manager)
+
+    @ast: parsed pycparser.c_ast object
+    @types_mngr: types manager
+    """
+
+    if types_mngr.layout == types_mngr.LAYOUT_COMPIL:
+        return ast_gen_union_align_size_compil(ast, types_mngr)
+    elif types_mngr.layout == types_mngr.LAYOUT_PACKED:
+        return ast_gen_union_align_size_packed(ast, types_mngr)
+    else:
+        raise RuntimeError("Unknown layout %r" % types_mngr.layout)
+
+
+def ast_gen_obj_align_size(ast, types_mngr):
+    """Evaluates the size/alignment of a C ast struct/union
+    (packed style in type manager)
+
+    @ast: parsed pycparser.c_ast object
+    @types_mngr: types manager
+    """
+
+    if isinstance(ast, c_ast.Struct):
+        obj = ast_gen_struct_align_size(ast, types_mngr)
+    elif isinstance(ast, c_ast.Union):
+        obj = ast_gen_union_align_size(ast, types_mngr)
+    else:
+        raise NotImplementedError("Not implemented!")
+
+    fix_recursive_objects(types_mngr, obj)
+    return obj
+
+
+def ast_parse_declarations(ast, types_mngr):
```

Comment:
The code is modified, but not as you suggested:
Those function became members of `type_mngr`. So layout disappears, and is handled in two subclasses.